packages/ceramic-core/src/dispatcher.ts

```diff
@@ -34,6 +37,15 @@ export default class Dispatcher extends EventEmitter {
   }
 
   async storeRecord (data: any): Promise<CID> {
+    if (DoctypeUtils.isSignedRecordDTO(data)) {
+      const { jws, linkedBlock } = data
+      // put the JWS into the ipfs dag
+      const cid = await this._ipfs.dag.put(jws, { format: 'dag-jose', hashAlg: 'sha2-256' })
+      // put the payload into the ipfs dag
+      await this._ipfs.block.put(linkedBlock, { cid: jws.link })
+      Object.assign(data, await this.retrieveRecord(cid)) // reload record
```

Comment:
Not sure I follow why this line is needed. Also why do you assign the dag-jose content to the `data` variable? Seems like this could have unforseen side effects?

Reply:
I removed that from `Dispatcher` and handled it elsewhere. The thing that happens is the following:

Initial state: `rec = { jws: {}, linkedBlock: {} }`

- `dag.put` `jws`
- `block.put` `linkedBlock`
- `dag.get` `rec.jws.link` throws an error

Reply:
Really? What error does it throw?

Reply:
That's the issue that I pasted yesterday. That's why I bypassed it like that.

The stacktrace:

```
TypeError: string.startsWith is not a function
        at toCidAndPath (/home/circleci/project/node_modules/ipfs-core-utils/src/to-cid-and-path.js:24:14)
        at fn (/home/circleci/project/node_modules/ipfs/src/core/components/dag/get.js:13:9)
        at Object.get (/home/circleci/project/node_modules/ipfs/src/core/utils.js:153:46)
        at Dispatcher.<anonymous> (/home/circleci/project/packages/ceramic-core/src/dispatcher.ts:52:44)
        at Generator.next (<anonymous>)
        at /home/circleci/project/packages/ceramic-core/lib/dispatcher.js:31:67
        at new Promise (<anonymous>)
        at __awaiter (/home/circleci/project/packages/ceramic-core/lib/dispatcher.js:10:10)
        at Dispatcher.retrieveRecord (/home/circleci/project/packages/ceramic-core/lib/dispatcher.js:110:12)
        at Ceramic.<anonymous> (/home/circleci/project/packages/ceramic-core/src/ceramic.ts:270:45)
```

Reply:
oh, could it be that `jws.link` is a string instead of a CID for whatever reason? Btw, still don't get how retrieving the record and assigning it to the data object would help.

Reply:
It's definitely a CID. I'm not assigning it to `data` in the `Dispatcher` anymore since that can introduce potential issues in the future.

This can be easily reproduced, just replace `_createDocFromGenesis` with:

```ts
async _createDocFromGenesis(genesis: any, opts: DocOpts = {}): Promise<Document> {
    const genesisCid = await this.dispatcher.storeRecord(genesis)
    let doc = this.getDocFromMap(genesisCid)
    if (doc) {
      return doc
    }

    let doctypeHandler
    if (DoctypeUtils.isSignedRecordDTO(genesis)) {
      const payload = await this.dispatcher.retrieveRecord(genesis.jws.link)
      doctypeHandler = this.findHandler(payload)
    } else {
      doctypeHandler = this.findHandler(genesis)
    }

    doc = await Document.create(genesisCid, doctypeHandler, this.dispatcher, this.pinStore, this.context, opts, this._validateDocs);
    const normalizedId = DoctypeUtils.normalizeDocId(doc.id)
    this._docmap[normalizedId] = doc
    return doc
  }
```

Maybe I'm missing something but this should work, right?

Reply:
That makes sense to me 👍

Reply:
This is the change that I made in order to work:

```ts
async _createDocFromGenesis(genesis: any, opts: DocOpts = {}): Promise<Document> {
    const genesisCid = await this.dispatcher.storeRecord(genesis)
    let doc = this.getDocFromMap(genesisCid)
    if (doc) {
      return doc
    }

    let doctypeHandler
    genesis = await this.dispatcher.retrieveRecord(genesisCid)
    if (DoctypeUtils.isSignedRecord(genesis)) {
      const payload = await this.dispatcher.retrieveRecord(genesis.link)
      doctypeHandler = this.findHandler(payload)
    } else {
      doctypeHandler = this.findHandler(genesis)
    }

    doc = await Document.create(genesisCid, doctypeHandler, this.dispatcher, this.pinStore, this.context, opts, this._validateDocs);
    const normalizedId = DoctypeUtils.normalizeDocId(doc.id)
    this._docmap[normalizedId] = doc
    return doc
  }
```